heka/_service.sls

```diff
@@ -0,0 +1,269 @@
+  - user: root
+  - mode: 644
+  - template: jinja
+
+{%- else %}
+
+heka_{{ service_name }}_service_file:
+  file.managed:
+  - name: /etc/init/{{ service_name }}.conf
+  - user: root
+  - mode: 644
+  - template: jinja
+
+heka_{{ service_name }}_service_wrapper:
+  file.managed:
+  - name: /usr/local/bin/{{ service_name }}.log
+  - user: root
+  - mode: 755
+  - template: jinja
+
+{%- endif %}
+
+
+{# Setup basic structure for all roles so updates can apply #}
+
+{%- set service_grains = {
+  'heka': {
+    'log_collector': {
+      'decoder': {},
+      'input': {},
+      'filter': {}
+    },
+    'metric_collector': {
+      'decoder': {},
+      'input': {},
+      'filter': {}
+    },
+    'remote_collector': {
+      'decoder': {},
+      'input': {},
+      'filter': {}
+    },
+    'aggregator': {
+      'decoder': {},
+      'input': {},
+      'filter': {}
+    }
+  }
+} %}
+
+
+{# Loading the other services' support metadata for local roles #}
+
+{%- if service_name in ['log_collector', 'metric_collector'] %}
+
+{%- for service_name, service in pillar.items() %}
+{%- if service.get('_support', {}).get('heka', {}).get('enabled', False) %}
+
+{%- set grains_fragment_file = service_name+'/meta/heka.yml' %}
+{%- set grains_yaml = load_grains_file(grains_fragment_file)|load_yaml %}
+{%- set _dummy = service_grains.heka.update(grains_yaml) %}
+
+{%- endif %}
+{%- endfor %}
+
+{%- endif %}
+
+
+{# Loading the other services' support metadata from salt-mine #}
+
+{%- if service_name in ['remote_collector', 'aggregator'] %}
+
+{%- for node_name, node_grains in salt['mine.get']('*', 'grains.items').iteritems() %}
+{%- if node_grains.heka is defined %}
+
+{%- set _dummy = service_grains.heka.update(node_grains.heka) %}
+
+{%- endif %}
+{%- endfor %}
+
+{%- endif %}
+
+
+{# Overriding aggregated metadata from user-space pillar data #}
+
+{%- for service_grain_name, service_grain in service_grains.items() %}
+{% if salt['pillar.get']('heka:'+service_grain_name) %}
+
+{%- for service_action_name, service_action in service_grain.items() %}
+{%- if salt['pillar.get']('heka:'+service_grain_name).get(service_action_name, False) is mapping %}
+{%- set _dummy = service_grains.heka.[service_grain_name].[service_action_name].update(salt['pillar.get']('heka:'+service_grain_name+':'+service_action_name)) %}
+{%- endif %}
+{%- endfor %}
+
+{%- endif %}
+{%- endfor %}
+
+
+/etc/{{ service_name }}/global.toml:
+  file.managed:
+  - source: salt://heka/files/toml/global.toml
+  - template: jinja
+  - mode: 640
+  - group: heka
+  - defaults:
+    service_name: {{ service_name }}
+  - require:
+    - file: heka_{{ service_name }}_conf_dir
+  - require_in:
+    - file: heka_{{ service_name }}_conf_dir_clean
+  - watch_in:
+    - service: heka_{{ service_name }}_service
+
+{%- for decoder_name, decoder in service_grains.heka.[service_name].decoder.iteritems() %}
+
+/etc/{{ service_name }}/10-decoder-{{ decoder_name }}-{{ decoder.engine }}.toml:
```

Comment:
You could maybe use subdirectory for decoders to avoid having directory with tons of files.